clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7430,6 +7430,212 @@ static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
+  QualType ValType = ValArg->getType();
+  if (isAddIRAttributesValidStringType(ValType) || ValType->isNullPtrType() ||
+      ValType->isIntegerType() || ValType->isFloatingType() ||
+      ValType->isCharType() || ValType->isEnumeralType())
+    return false;
+
+  return S->Diag(ValArg->getBeginLoc(),
+                 diag::err_sycl_add_ir_attribute_invalid_value)
+         << CI;
+}
+
+// Checks and evaluates arguments of an add_ir_attributes_* attribute. Returns
+// true if an error occured.
+static bool evaluateAddIRAttributesArgs(Expr **Args, size_t ArgsSize, Sema *S,
+                                        const AttributeCommonInfo &CI) {
+  ASTContext &Context = S->getASTContext();
+
+  // Check filter list if it is the first argument.
+  bool HasFilter = ArgsSize && isa<InitListExpr>(Args[0]);
+  if (HasFilter && checkAddIRAttributesFilterListExpr(Args[0], S, CI))
+    return true;
+
+  llvm::SmallVector<PartialDiagnosticAt, 8> Notes;
+  bool HasDependentArg = false;
+  for (unsigned I = HasFilter; I < ArgsSize; I++) {
+    Expr *&E = Args[I];
+
+    if (isa<InitListExpr>(E))
+      return S->Diag(E->getBeginLoc(),
+                     diag::err_sycl_add_ir_attr_filter_list_invalid_arg)
+             << CI;
+
+    if (E->isValueDependent() || E->isTypeDependent()) {
+      HasDependentArg = true;
+      continue;
+    }
+
+    Expr::EvalResult Eval;
+    Eval.Diag = &Notes;
+    bool Result = E->EvaluateAsConstantExpr(Eval, Context);
+    if (!Result || !Notes.empty()) {
+      S->Diag(E->getBeginLoc(), diag::err_attribute_argument_n_type)
+          << CI << (I + 1) << AANT_ArgumentConstantExpr;
+      for (auto &Note : Notes)
+        S->Diag(Note.first, Note.second);
+      return true;
+    }
+    assert(Eval.Val.hasValue());
+    E = ConstantExpr::Create(Context, E, Eval.Val);
+  }
+
+  // If there are no dependent expressions, check for expected number of args.
+  if (!HasDependentArg && ArgsSize && (ArgsSize - HasFilter) & 1)
+    return S->Diag(CI.getLoc(), diag::err_sycl_add_ir_attribute_must_have_pairs)
+           << CI;
+
+  // If there are no dependent expressions, check argument types.
+  // First half of the arguments are names, the second half are values.
+  unsigned MidArg = (ArgsSize - HasFilter) / 2 + HasFilter;
+  if (!HasDependentArg)
+    for (unsigned I = HasFilter; I < ArgsSize; ++I)
+      if ((I < MidArg && checkAddIRAttributesNameExpr(Args[I], S, CI)) ||
+          (I >= MidArg && checkAddIRAttributesValueExpr(Args[I], S, CI)))
+        return true;
+  return false;
+}
+
+SYCLAddIRAttributesFunctionAttr *Sema::MergeSYCLAddIRAttributesFunctionAttr(
+    Decl *D, const SYCLAddIRAttributesFunctionAttr &A) {
+  if (const auto *ExistingAttr =
+          D->getAttr<SYCLAddIRAttributesFunctionAttr>()) {
+    Diag(ExistingAttr->getLoc(), diag::warn_duplicate_attribute_exact) << &A;
+    Diag(A.getLoc(), diag::note_previous_attribute);
+    return nullptr;
+  }
```

Comment:
This seems surprising to me. We should not be diagnosing a case like this:
```
// Header file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void f();
// Source file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void f() { ... }
```
This code is eminently reasonable (especially if the attribute is hidden behind a macro). Why are we being so strict here instead of merging lists in a way the user may likely expect?

Reply:
There's a few approaches we could take, and it's really up to the designers as to which one to take. The easiest solution that fits most closely with what I *think* the intent is would be to only diagnose if the lists differ, but silently accept them when the lists are the same. It's still not completely cut and dry though, considering:
```
// Header file
[[__sycl_detail__::add_ir_attributes_function("Attr2", "Attr1", true, 1)]] void f();
// Source file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void f() { ... }
```
These are functionally identical (I believe) but we might still want to reject it because they're not identical. However, we probably don't want to take that to the logical extreme either:
```
#define FROBBLE true

// Header file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void f();
// Source file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, FROBBLE)]] void f() { ... }
```
I see no reason why we'd want to reject this one despite them not being identical. I'm very much on the fence about this example:
```
// Header file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void f();
// Source file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1 + 0, true && true)]] void f() { ... }
```
As this is also identical, but a bit less identical than the previous examples.

Reply:
Here's some future-looking direction that might help here.  There is a proposed extension for kernel properties that defines a new macro named `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY`:

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_properties.asciidoc#device-functions

That macro will probably be implemented by expanding to `[[__sycl_detail__::add_ir_attributes_function()]]`.  This means that we will want to allow multiple occurrences of that attribute on the same function declaration.  The expectation is that the set of IR attributes on the function should be the *union* of all the parameters from all these C++ attributes.

The  proposed API spec does not really say whether it is legal to use the macro on *both* the declaration and the definition of the function like:

```
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::device_has<aspect::fp16, aspect::atomic64>))
void foo();

SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::device_has<aspect::fp16, aspect::atomic64>))
void foo() {/*definition*/}
```

I suspect that we probably do want to allow this (but not require it).  Tagging @Pennycook to see what he thinks.

Reply:
Thanks, that's helpful!

Attributes which are inheritable (like all of these) are automatically inherited onto subsequent redeclarations, and a definition is itself a declaration. So I think we definitely want to allow (but not require) the attribute to be written on the declaration and the definition.

Taking the union of these sets is interesting as we now need to make sure we are looking at the actual values of both the LLVM IR attribute names and values so that we unique properly. e.g., 
```
// Header file
[[__sycl_detail__::add_ir_attributes_function("Attr2", "Attr1", true, 1)]] void f();
// Source file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void f() { ... }
```
These should be treated as the same thing. It does still leave some question as to how equivalent the tokens need to be though -- because these are constant expressions, I *think* these should also be valid and treated as the same thing:
```
// Header file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void f();
// Source file
[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1 + 0, true && true)]] void f() { ... }
```

Reply:
I think the definition of our macro `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY` will prevent cases like `1 + 0` or `true && true` from ever occurring.  From an aesthetic point of view, though, it does seem better to check for uniqueness after `constexpr` expressions are evaluated.

Reply:
@gmlueck - What should be the expected behavior for redeclarations with different filter lists? For example,

```c++
// Header file
[[__sycl_detail__::add_ir_attributes_function({"Attr1", "Attr3"}, "Attr1", "Attr2", 1, true)]] void f();
// Source file
[[__sycl_detail__::add_ir_attributes_function({"Attr1"}, "Attr3", "Attr2", 1, true)]] void f() { ... }
```
Should the merged filter be the union, intersection, or disallowed? If it should be the union, would that mean the resulting IR should have both "Attr1" and "Attr3"?

Reply:
Apologies for the slow response here.  I thought I responded yesterday, but I think I forgot to actually press the "Comment" button...

I agree that we want to allow the attributes in both places, but the union behavior should only apply if there are multiple instances of the attribute on the same definition/declaration.  A function can only be associated with one set of properties, so providing different properties on the definition and declaration should be disallowed.

I'll update the kernel properties spec to clarify this.

Reply:
@steffenlarsen

The only time we will currently have redeclarations with `[[__sycl_detail__::add_ir_attributes_function()]]` will be via `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY`, and that macro will not use the filter parameter.  Therefore, I think your situation will not occur in practice.

A sensible semantic might be like this, though.  Apply the filter function to each C++ attribute individually.  Then take the union of the post-filtered IR attributes across all the C++ attributes on the same declaration.  This yields a set of IR attributes attached to that declaration.  If there is a redeclaration of the same function in the same TU and the redeclaration also has `[[__sycl_detail__::add_ir_attributes_function()]]`, then the IR attributes on that redeclaration (post-filtered and unionized) must exactly match the IR attributes on the other declaration.

BTW, I was thinking that the property names listed in the filter must all exist in that same C++ attribute.  Therefore, this example would be illegal because `Attr3` is in the filter list, but not in the main parameter list of the attribute:

```
[[__sycl_detail__::add_ir_attributes_function({"Attr1", "Attr3"}, "Attr1", "Attr2", 1, true)]] void f();
```

Reply:
> BTW, I was thinking that the property names listed in the filter must all exist in that same C++ attribute. Therefore, this example would be illegal because Attr3 is in the filter list, but not in the main parameter list of the attribute:

I was under the impression that the intention of the filter was to ignore property meta-names that do not need IR attributes for a given definition. If that is the case, requiring all filters in the filter list to also appear in the attribute would mean the property lists would always need the corresponding properties.

Reply:
> I agree that we want to allow the attributes in both places, but the union behavior should only apply if there are multiple instances of the attribute on the same definition/declaration. A function can only be associated with one set of properties, so providing different properties on the definition and declaration should be disallowed.

I think there may be some confusion. What I hear you saying is:
```
[[attr(arg1, arg2), attr(arg3, arg4)]][[attr(arg5)]] void func1();
void func1() {} // OK

[[attr(arg1, arg2), attr(arg3, arg4)]][[attr(arg5)]] void func2();
[[attr(arg1, arg2, arg3, arg4, arg5)]] void func2() {} // OK?

[[attr(arg1, arg2)]] void func3();
[[attr(arg3, arg4)]] void func3(); // Error
[[attr(arg5)]] void func3() {}  // Error
```
However, that's not really how attributes work in practice. When an attribute is "inheritable", it gets automatically replicated on redeclarations (and a definition is a declaration). e.g., https://godbolt.org/z/eh1a5qs71 (see the AST dump and note how the second declaration as two attributes on it). When an attribute is not inheritable, it does not get replicated on redeclarations and that's typically done for very specific reasons (like disallowing the attribute from appearing on a declaration which is not a definition, etc). Not inheriting the attribute on subsequent redeclarations means declaration lookup behavior is critical for the use of the attribute -- if lookup finds the "wrong" declaration, it will be missing the attributes. This means you can run into a situation like:
```
[[attr]] void func(); 

void use() {
  func(); // Sees attributes
}

void func() {}

void use_again() {
  func(); // Sees no attributes
}
```
which is why almost no attributes work the way you're proposing.

Reply:
What is your proposed behavior in a case like this, @AaronBallman?

```
[[attr(arg1)]] [[attr(arg2)]] void func();

void use1() {
  func();  // Which attributes does func have here?
}

[[attr(arg3)]] void func();

void use2() {
  func();  // Which attributes does func have here?
}
```

@Pennycook was proposing that the second declaration of `func()` should raise an error, but I think you are saying that it should be allowed.  If that is the case, which attributes does `func()` have at the two call sites?

Reply:
I'm definitely confused -- I don't know very much about how attributes work, so thanks for your patience here and for the explanation.

I hadn't actually considered cases like this:
```c++
[[attr(arg1, arg2), attr(arg3, arg4)]][[attr(arg5)]] void func2();
[[attr(arg1, arg2, arg3, arg4, arg5)]] void func2() {} // OK?
```
I wouldn't actually have expected this to work!  I think it's also very unlikely for code like this to exist unless developers start playing with implementation details.  Because the extension uses a macro (yup) and each macro can only accept one property, I think the only legal form is:
```c++
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::arg1)) // NB: Lazily using sycl:: here to denote a property class
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::arg2))
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::arg3))
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::arg4))
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::arg5))
void func2();
```
...expanding to something like:
```c++
[[attr(arg1)]][[attr(arg2)]][[attr(arg3)]][[attr(arg4)]][[attr(arg5)]] void func2();
```

I'm still confused about why the inheritance behavior causes a problem.  For your godbolt example, couldn't we just have a pass that checks for two `FunctionDecl`s with different attributes, and throw an error then?

I think it's very important that each function has exactly one set of properties associated with it, and ideally we'd be able to diagnose cases where that isn't true.  Some of the use-cases we have in mind for these sorts of function properties are things like enforcing a particular SIMD size, and if we allow users to specify multiple SIMD sizes for a single function then compilation is likely to fail in a very nasty way later on (maybe even as late as device compilation).

Is there a good way to design an attribute that will give the right error behavior?  Or do we just need to say that specifying anything other than exactly the same set of attributes is UB, and that the best way to avoid nasty surprises is to only specify the attributes on the declaration?

Reply:
@gmlueck -- answer a bit unclear, to be honest. The semantic attributes at Sema time I would expect `use1()` would see two attributes, and `use2()` would see three attributes. At CodeGen time, when we're lowering things to LLVM IR, we have more flexibility and could go either way. In terms of these specific attributes, I'm not certain what the best approach is.

My proposal is: for the moment, accept identical attribute lists on a redeclaration, reject different ones. In the future, we can relax the restriction and nail down the union behavior. e.g.,
```
[[attr1]][[attr2]] void func();
[[attr1, attr2]] void func(); // OK
[[attr1, attr2]] void func() { } // OK

[[attr1, attr2]] void func2();
[[attr3]] void func2(); // Error

[[attr1]] void func3();
[[attr1]][[attr2]] void func3(); // Error
```
This leaves one open question:
```
[[attr1, attr2]] void func();
void func() {} // Ok?
```
I think we want this to be OK (so the `func` definition inherits from the prior declaration), but that kind of suggests that we're merging attributes already.

Reply:
>  Because the extension uses a macro (yup) and each macro can only accept one property, I think the only legal form is:

The trouble is -- as far as the compiler is concerned, the macro is irrelevant because users can write the attributes directly. Also, because these are intended to be general attributes for blasting LLVM IR attributes onto things, so the macro you're using today isn't necessarily the same as other macros you add in the future to use the same general mechanisms.

> I'm still confused about why the inheritance behavior causes a problem. For your godbolt example, couldn't we just have a pass that checks for two FunctionDecls with different attributes, and throw an error then?

It doesn't cause a problem per se; we just don't know what the design intent is because we could either merge the attribute lists, or we could require them to be identical/not present, etc. So we could check for different attributes when we do function declaration merging (I had suggested that just above before noticing you had replied in the meantime).

> Is there a good way to design an attribute that will give the right error behavior?

I had suggested this above: My proposal is: for the moment, accept identical attribute lists on a redeclaration, reject different ones. In the future, we can relax the restriction and nail down the union behavior. Do you think something like that would be a good path forward?

Reply:
> This leaves one open question:
> 
> ```c++
> [[attr1, attr2]] void func();
> void func() {} // Ok?
> ```
> I think we want this to be OK (so the func definition inherits from the prior declaration), but that kind of suggests that we're merging attributes already.

I agree that it's desirable for this to be OK.  I expect developers will want to put the properties in headers and then omit them later (so that they can update the properties without having to change multiple files).


> My proposal is: for the moment, accept identical attribute lists on a redeclaration, reject different ones. In the future, we can relax the restriction and nail down the union behavior. Do you think something like that would be a good path forward?

This sounds good to me.

For the use-cases I'm worried about, I can't imagine we'll ever want to relax the restriction.  (Unless you consider attributes on the declaration and no attributes on the definition, as in your open question, to be a relaxation here.)  I can see what you mean about future features trying to generate this attribute and potentially expecting different behavior, though.  I think it makes sense to revisit the union behavior once we know what those use-cases would expect.

Reply:
I also think your proposal sounds good, @AaronBallman.  And I agree with @Pennycook that the following should be allowed:

```
[[attr1, attr2]] void func();
void func() {} // Ok?
```

Reply:
Excellent, thank you both!

@steffenlarsen, it seems we've got consensus here. Do you have a good handle on how to approach this? (If not, let me know and I can give some hints.)

Reply:
I have added tests to make sure that filters also are considered when comparing attributes.